diff -urN qe-7.2_ori/PP/src/punch_plot.f90 qe-7.2/PP/src/punch_plot.f90
--- qe-7.2_ori/PP/src/punch_plot.f90	2023-03-27 17:12:09.000000000 +0100
+++ qe-7.2/PP/src/punch_plot.f90	2023-10-08 01:04:01.000000000 +0100
@@ -199,7 +199,10 @@
      CALL v_h( rho%of_g(:,1), ehart, charge, raux2 )
 
      raux(:) = raux2(:,1)
-     IF (tefield.and.dipfield) CALL add_efield(raux, dummy, rho%of_r(:,1),.true.)
+     !IF (tefield.and.dipfield) CALL add_efield(raux, dummy, rho%of_r(:,1),.true.)
+     IF (tefield.and.dipfield) THEN
+        WRITE(stdout,*) "V_hartree and V_bare is shown but V_saw is printed in pw by jay"
+     ENDIF
      
      DEALLOCATE( raux2 )
 
