diff -urN qe-6.3/Modules/input_parameters.f90 qe-6.3_ces/Modules/input_parameters.f90
--- qe-6.3/Modules/input_parameters.f90	2018-07-07 00:36:48.000000000 +0900
+++ qe-6.3_ces/Modules/input_parameters.f90	2022-04-15 12:48:36.549370121 +0900
@@ -272,6 +272,15 @@
 
           ! if .TRUE., perform exact exchange calculation using Wannier functions (X. Wu et al., Phys. Rev. B. 79, 085102 (2009))
 
+! DFT-CES start
+        LOGICAL :: dft_ces = .false.
+        ! If true reads MD rho from file "file_pot"
+        CHARACTER(len=80) :: rho_ces = './rho_ces.cube'
+        CHARACTER(len=80) :: rho_ces_ion = './rho_ces_ion.cube'
+        CHARACTER(len=80) :: rho_ces_rep = './rho_ces_rep.cube'
+        ! File containing the MD.rho as gaussian cube format
+        ! with conventional charge sign: -1 = electron
+! DFT-CES end
         LOGICAL  :: lfcpopt = .FALSE. ! FCP optimisation switch
         LOGICAL  :: lfcpdyn = .FALSE. ! FCP thermostat enabled if .true.
         !
@@ -285,6 +294,9 @@
           gdir, nppstr, wf_collect, lelfield, nberrycyc, refg,            &
           tefield2, saverho, tabps, lkpoint_dir, use_wannier, lecrpa,     &
           tqmmm, vdw_table_name, lorbm, memory, point_label_type,         &
+! DFT-CES start
+          dft_ces, rho_ces, rho_ces_ion, rho_ces_rep,                    &
+! DFT-CES end
           lfcpopt, lfcpdyn, input_xml_schema_file, gate                                        
 !
 !=----------------------------------------------------------------------------=!
diff -urN qe-6.3/Modules/read_namelists.f90 qe-6.3_ces/Modules/read_namelists.f90
--- qe-6.3/Modules/read_namelists.f90	2018-07-07 00:36:48.000000000 +0900
+++ qe-6.3_ces/Modules/read_namelists.f90	2022-04-15 12:48:36.631370124 +0900
@@ -130,6 +130,12 @@
        lfcpopt = .FALSE.
        lfcpdyn = .FALSE.
        !
+! DFT-CES start
+       dft_ces = .FALSE.
+       rho_ces = './rho_ces.cube'
+       rho_ces_ion = './rho_ces_ion.cube'
+       rho_ces_rep = './rho_ces_rep.cube'
+! DFT-CES end
        CALL get_environment_variable( 'QEXML', input_xml_schema_file )
        !
        RETURN
@@ -740,6 +746,12 @@
        CALL mp_bcast( lfcpdyn,       ionode_id, intra_image_comm )
        CALL mp_bcast( input_xml_schema_file, ionode_id, intra_image_comm )
        CALL mp_bcast( gate,          ionode_id, intra_image_comm ) !TB
+! DFT-CES start
+       CALL mp_bcast( dft_ces,       ionode_id, intra_image_comm )
+       CALL mp_bcast( rho_ces,       ionode_id, intra_image_comm )
+       CALL mp_bcast( rho_ces_ion,   ionode_id, intra_image_comm )
+       CALL mp_bcast( rho_ces_rep,   ionode_id, intra_image_comm )
+! DFT-CES end
        !
        RETURN
        !
